File: kotlin-src/src/main/kotlin/com/memetoclasm/districtlive/ingestion/enrichment/ArtistEnricher.kt
package com.memetoclasm.districtlive.ingestion.enrichment

/**
 * Port interface implemented by each external enrichment adapter.
 * Accepts a cleaned artist name; returns an EnrichmentResult or null if no
 * confident match was found. Null is a normal result (not an error).
 */
interface ArtistEnricher {
    val source: EnrichmentSource
    suspend fun enrich(cleanedName: String): EnrichmentResult?
}

/**
 * Identifies which external system produced an EnrichmentResult.
 */
enum class EnrichmentSource {
    MUSIC_BRAINZ,
    SPOTIFY,
    OLLAMA
}

/**
 * Data returned by an ArtistEnricher adapter.
 * Fields not populated by the adapter are null / empty.
 *
 * @param canonicalName Better-spelled name from the external source, or null if no improvement found.
 * @param externalId    Stable external identifier (MB UUID or Spotify artist ID).
 * @param tags          Genre/tag list from the source (mb_tags OR spotify_genres — never mixed).
 * @param imageUrl      Artist image URL from the source.
 * @param confidence    Match confidence: Jaro-Winkler score for MusicBrainz, 1.0 for Spotify.
 */
data class EnrichmentResult(
    val canonicalName: String?,
    val externalId: String?,
    val tags: List<String>,
    val imageUrl: String?,
    val confidence: Double
)

File: src/ports/artist_enricher.rs
use async_trait::async_trait;

use crate::domain::{
    artist::{EnrichmentResult, EnrichmentSource},
    error::EnrichmentError,
};

#[async_trait]
pub trait ArtistEnricher: Send + Sync {
    /// Which enrichment service this adapter queries.
    fn source(&self) -> EnrichmentSource;

    /// Look up metadata for an artist by name.
    ///
    /// Returns `Ok(None)` if no match is found (not an error — just unknown artist).
    /// Returns `Err` only for transient failures (HTTP errors, rate limits).
    async fn enrich(
        &self,
        name: &str,
    ) -> Result<Option<EnrichmentResult>, EnrichmentError>;
}
